src/librustc_mir/borrow_check/nll/region_infer/error_reporting/outlives_suggestion.rs

```diff
@@ -0,0 +1,428 @@
+use syntax::errors::Applicability;
+use syntax_pos::{BytePos, Pos, Span, DUMMY_SP};
+
+use super::region_name::RegionName;
+use borrow_check::nll::region_infer::RegionInferenceContext;
+
+/// Collects information about outlives constraints that needed to be added for a given MIR node
+/// corresponding to a function definition.
+///
+/// Adds a suggestion to outlives errors:
+///
+/// ```text
+/// suggestion: try adding the following outlives constraint
+///    |
+/// LL |     fn foo<'a, 'b>(&'a u32) -> &'b u32 where 'a: 'b {
+///    |                                        ^^^^^^^^^^^^
+/// ```
+pub struct OutlivesSuggestionBuilder {
+    /// The MIR DefId of the fn with the lifetime error.
+    mir_def_id: DefId,
+
+    /// The list of outlives constraints that need to be added. Specifically, we map each free
+    /// region to all other regions that it must outlive. I will use the shorthand `fr:
+    /// outlived_frs`. Not all of these regions will already have names necessarily. Some could be
+    /// implicit free regions that we inferred. These will need to be given names in the final
+    /// suggestion message.
+    constraints_to_add: FxHashMap<RegionVid, Vec<RegionVid>>,
+
+    /// A list of lifetimes that don't have user-defined names. These will be used in the "name
+    /// these lifetimes" portion of the suggestion.
+    names_to_add: FxHashMap<RegionVid, String>,
+
+    new_name_counter: usize,
+}
+
+impl OutlivesSuggestionBuilder {
+    /// Create a new builder for the given MIR node representing a fn definition.
+    pub fn new(mir_def_id: DefId) -> Self {
+        OutlivesSuggestionBuilder {
+            mir_def_id,
+            constraints_to_add: FxHashMap::default(),
+            names_to_add: FxHashMap::default(),
+            new_name_counter: 0,
+        }
+    }
+
+    /// Add the outlives constraint `fr: outlived_fr` to the set of constraints we need to suggest.
+    pub fn collect_constraint(
+        &mut self,
+        fr: RegionVid,
+        outlived_fr: RegionVid,
+        rinfcx: &RegionInferenceContext<'tcx>,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+    ) {
+        debug!("Collected {:?}: {:?}", fr, outlived_fr);
+
+        self.lookup_or_create_region_name(fr, rinfcx, tcx);
+        self.lookup_or_create_region_name(outlived_fr, rinfcx, tcx);
+
+        if let Some(ref mut outlived_frs) = self.constraints_to_add.get_mut(&fr) {
+            outlived_frs.push(outlived_fr);
+        } else {
+            self.constraints_to_add.insert(fr, vec![outlived_fr]);
+        }
+    }
+
+    /// Either get the existing user-define name of the given region or create one.
+    fn lookup_or_create_region_name(
+        &mut self,
+        fr: RegionVid,
+        rinfcx: &RegionInferenceContext<'tcx>,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+    ) {
+        /// All generated names will be of the form `'PREFIXN`, where `N` is a unique integer.
+        const PREFIX: &str = "generatedl";
```

Comment:
@estebank What is a better way to generate a fresh unused name?

Reply:
The NLL code that creates unused names is https://github.com/rust-lang/rust/blob/master/src/librustc_mir/borrow_check/nll/region_infer/error_reporting/region_name.rs, but right now that only uses numbers to refer to unnamed lifetimes. I know that there was other code to create a sequence of named lifetimes that could be copy-pasted into code without being a syntax error (I think it was the g, h, i, j sequence), but don't recall where it was present.